cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -127,14 +184,27 @@ public ModelCompletenessRequirements clusterModelCompletenessRequirements() {
   protected boolean selfSatisfied(ClusterModel clusterModel, BalancingAction action) {
     Broker destinationBroker = clusterModel.broker(action.destinationBrokerId());
     Replica sourceReplica = clusterModel.broker(action.sourceBrokerId()).replica(action.topicPartition());
-    // If the source broker is dead and currently self healing dead brokers only, then the action must be executed.
+    // If the source broker is dead and currently self healing dead brokers only, unless it is replica swap, the action
+    // must be executed.
     if (!sourceReplica.broker().isAlive() && _selfHealingDeadBrokersOnly) {
-      return true;
+      return action.balancingAction() != REPLICA_SWAP;
     }
 
-    //Check that current destination would not become more unbalanced.
-    return isLoadUnderBalanceUpperLimitAfterChange(clusterModel, sourceReplica.load(), destinationBroker, ADD) &&
-        isLoadAboveBalanceLowerLimitAfterChange(clusterModel, sourceReplica.load(), sourceReplica.broker(), REMOVE);
+    switch (action.balancingAction()) {
+      case REPLICA_SWAP:
+        Replica destinationReplica = destinationBroker.replica(action.destinationTopicPartition());
+        double sourceUtilizationDelta = destinationReplica.load().expectedUtilizationFor(resource())
+                                        - sourceReplica.load().expectedUtilizationFor(resource());
+
+        return sourceUtilizationDelta != 0 && isLoadStillWithinLimitsAfterSwap(sourceReplica, destinationReplica);
```

Comment:
Not sure if this is right. If the upper limit is 30, lower limit is 10. source broker has utilization of 40 and destination broker has utilization of 3, when source delta is -5, are we going to reject the proposal?

Reply:
We do _not_ reject the example above -- this method accepts it.
`isLoadStillWithinLimitsAfterSwap` does the same task as `isLoadUnderBalanceUpperLimitAfterChange` and `isLoadAboveBalanceLowerLimitAfterChange` for (1) the source and (2) destination brokers. In other words, it checks whether the load is above the upper limit for the broker after it would receive more load and checks whether the load is under the lower limit after the broker would gave away its load.

Would the change `isLoadStillWithinLimitsAfterSwap` --> `!isSwapBoostingLoadLimitViolation` make this functionality more clear?